# Part 3: More Data Structures

In the previous notebook, we looked at ways to store and work with a collection of data values within a sequence, like a list or an array. We also explored insights we can gain from computing summary statistics on those collections of data. Now we need a way to store and work with **multiple** separate collections of information about the same population - e.g. in addition to the test scores of the class, perhaps we also know the students' IDs, and results from one other exam they have taken in the past. In this notebook, we will work with data structures that allow us to store multiple pieces of information, or **features**, about a population; and we will explore a powerful library that allows us to read in, explore and manipulate datasets.

## Table of Contents
   
**1. [Collections of Multiple Features](#multFeat)**   
&ensp;&ensp;&ensp;&ensp;**1.1.** [Dictionaries](#dict)  
&ensp;&ensp;&ensp;&ensp;**1.2.** [Matrices](#matrix)  
**2. [Pandas Library](#pd)**  
&ensp;&ensp;&ensp;&ensp;**2.1.** [Reading in Data](#import)  
&ensp;&ensp;&ensp;&ensp;**2.2.** [Exploring the Pandas DataFrame](#df)  
&ensp;&ensp;&ensp;&ensp;**2.3.** [Selecting Rows & Columns](#loc)  
&ensp;&ensp;&ensp;&ensp;**2.4.** [Applying Functions](#apply)  
&ensp;&ensp;&ensp;&ensp;**2.5.** [Data Aggregation](#group)  


---

## <u>1. Collections of Multiple Features</u><a id='multFeat'></a>

First, we will explore 2 important data structures that allow us to organize multiple features of a given population: dictionaries and matrices.

### 1.1. Dictionaries<a id='dict'></a>

We can use arrays to collect data points on a single feature of the population, like `test_scores`. When we have multiple arrays capturing different features of the same population, one way to store all of the different arrays is in a **dictionary**.

Here are two examples of a dictionary:

In [70]:
# ex1. dictionary of single values
numerals = {'I': 1, 'V': 5, "X": 10}
numerals

{'I': 1, 'V': 5, 'X': 10}

In [71]:
# ex2. dictionary of collections
class_dict = {"student_ID":np.arange(1, len(test_scores)+1), 
              "test_scores":test_scores}
class_dict

{'student_ID': array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15]),
 'test_scores': array([ 0. , 52.9, 69. , 72.1, 72.9, 73.3, 74.1, 77.1, 78.1, 78.2, 78.4,
        79.7, 80.2, 80.3, 82.2])}

<i class="fa fa-book" style="font-size:20px;"></i> &nbsp;**Definition:**
<div class="alert alert-success">
A <b>dictionary</b> organizes data into <b>key-value pairs</b>. This allows us to store and retrieve values indexed not by consecutive integers, but by descriptive keys. 

- <b>Keys:</b> Strings commonly serve as keys since they enable us to represent names of things. In the context of storing data, they are the column names, or names of the value(s) it represents. 
- <b>Values:</b> The data that we are storing. This can be a single value, or a collection of values. 
</div>

<br><i class="fa fa-thumb-tack" style="font-size:16px;"></i>&nbsp; **Accessing Dictionary Contents**  

1. Access a dictionary value by indexing the dictionary by the corresponding key:

In [72]:
# 1. get value associated with "test_scores"
class_dict['test_scores']

array([ 0. , 52.9, 69. , 72.1, 72.9, 73.3, 74.1, 77.1, 78.1, 78.2, 78.4,
       79.7, 80.2, 80.3, 82.2])

2. Dictionaries have methods that give us access to a list of its keys, values, and key-value pairs: 

In [73]:
## a. list of dictionary keys:
class_dict.keys()

dict_keys(['student_ID', 'test_scores'])

In [74]:
## b. list of dictionary values:
class_dict.values()

dict_values([array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15]), array([ 0. , 52.9, 69. , 72.1, 72.9, 73.3, 74.1, 77.1, 78.1, 78.2, 78.4,
       79.7, 80.2, 80.3, 82.2])])

In [75]:
## c. list of (key, value) pairs:
class_dict.items()

dict_items([('student_ID', array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15])), ('test_scores', array([ 0. , 52.9, 69. , 72.1, 72.9, 73.3, 74.1, 77.1, 78.1, 78.2, 78.4,
       79.7, 80.2, 80.3, 82.2]))])

<div class="alert alert-warning">
<i class="fa fa-info-circle" style="font-size:22px;color:orange"></i> &nbsp; Unlike lists and arrays, dictionary are <b>unordered</b> so the order in which the key:value pairs appear in the dictionary may change when you run code cells. 
</div>

<br><i class="fa fa-thumb-tack" style="font-size:16px;"></i>&nbsp; **Adding key-value Pairs** 

You can add a new item (key-value pair) into the exiting dictionary by assigning the value to a new name on the dictionary:

```python
dictionary['new_key'] = new_value
```

In [76]:
# adding a new entry
past_scores = np.array([89.0, 94.2, 78.0, 86.2, 81.2, 86.0, 88.3, 84.9, 88.1, 93.0, 82.2, 78.2, 96.1, 95.9, 98.2])

class_dict["past_test_score"] = past_scores
class_dict

{'student_ID': array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15]),
 'test_scores': array([ 0. , 52.9, 69. , 72.1, 72.9, 73.3, 74.1, 77.1, 78.1, 78.2, 78.4,
        79.7, 80.2, 80.3, 82.2]),
 'past_test_score': array([89. , 94.2, 78. , 86.2, 81.2, 86. , 88.3, 84.9, 88.1, 93. , 82.2,
        78.2, 96.1, 95.9, 98.2])}

In [77]:
class_dict.items()

dict_items([('student_ID', array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15])), ('test_scores', array([ 0. , 52.9, 69. , 72.1, 72.9, 73.3, 74.1, 77.1, 78.1, 78.2, 78.4,
       79.7, 80.2, 80.3, 82.2])), ('past_test_score', array([89. , 94.2, 78. , 86.2, 81.2, 86. , 88.3, 84.9, 88.1, 93. , 82.2,
       78.2, 96.1, 95.9, 98.2]))])

<div class="alert alert-warning">
<b>Note</b>: There can only be 1 value per key. If you attempt to assign a new value to the dictionary but specify a key name that already exists in the dictionary, the existing values associated with that key will be overwritten.
</div>

---

### 1.2. Matrices<a id='matrix'></a>

Dictionaries organize information by features - all data values capturing student IDs are boxed into one container and saved into a dictionary; and data values about test scores from last year are boxed up into a separate container and saved into the dictionary under a differet label.

But when you think about it, that is not the most helpful way to organize data when you are trying to **make predictions** about a specific case. For instance, say you were trying to guess what animal each record (row) is, given the following features:

||Opposable Thumbs|Class of Animal|Diet |Tail Length |Number of Legs | Flies|
|:-:|:-:|:-:|:-:|:-:|:-:|:-:|
|**0**|True|Mammal|Bananas| long | 2|False|
|**1**|False|Anthropod|Insects| none  |8| False|
|**2**|False|Bird|Fish|short|2|False|

We wouldn't want to look at the data one column at a time, the way dictionaries are organized, when we want to predict what animal record **0** might be. Instead, we'd want to look all the features of the one record at the same time:

||Opposable Thumbs|Animal Class|Diet |Tail Length | Wings |Number of Legs | Flies|
|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|
|**0**|True|Mammal|Bananas| long | False | 2|False|

To make a prediction about a particular record, we need to consider all its features - we want to organize information by record, not by column. This is exactly what a **matrix** is designed to do, and it is in this matrix form that we ultimately feed the data into machine learning models.

<i class="fa fa-book" style="font-size:20px;"></i> &nbsp;**Definition:**
<div class="alert alert-success">
    A <b>matrix</b> is a rectangular list*, or <b>a list of lists.</b> We say that matrix $M$ has shape $m \times n$:
    
* It has **m** rows: each row is a list of all features that describe a single record;
* It has **n** columns: each column is displayed as elements in the same position/index of every row, and represents a specific feature of the data

</div>

Our example above in matrix form would look like:

In [78]:
animal_matrix = [[ True,    'Mammal', 'Bananas',  'long', 2, False],
                 [False, 'Anthropod', 'Insects',  'none', 9, False],
                 [False,      'Bird',    'Fish', 'short', 2, False]]

animal_matrix

[[True, 'Mammal', 'Bananas', 'long', 2, False],
 [False, 'Anthropod', 'Insects', 'none', 9, False],
 [False, 'Bird', 'Fish', 'short', 2, False]]

Compare this to how the data would be represented in a dictionary:

In [79]:
animal_dict = {"Opposable Thumbs": [True, False, False],
                "Class of Animal": ['Mammal', 'Anthropod', 'Bird'],
                           "Diet": ['Bananas', 'Insects', 'Fish'],
                    "Tail Length": ['long', 'none', 'short'],
                 "Number of Legs": [2, 8, 2],
                          "Flies": [False, False, False]}
animal_dict

{'Opposable Thumbs': [True, False, False],
 'Class of Animal': ['Mammal', 'Anthropod', 'Bird'],
 'Diet': ['Bananas', 'Insects', 'Fish'],
 'Tail Length': ['long', 'none', 'short'],
 'Number of Legs': [2, 8, 2],
 'Flies': [False, False, False]}

<div class="alert alert-warning">
<i class="fa fa-info-circle" style="font-size:22px;color:orange"></i> &nbsp;*We use lists in this example to demonstrate what a matrix looks like, since the features are represented by different value types (and values in NumPy arrays must all be of the same type). However, NumPy's representation of the matrix <a href='https://numpy.org/doc/stable/reference/arrays.ndarray.html'><code>ndarray</code></a>, or the <b>n-dimensional array</b>, is usually preferred over using Python lists because NumPy arrays consume less memory and is able to handle operations much more efficiently than lists. Even though we have a mix of data types in our example, that does not mean we are stuck using lists. <b>There are many ways to transform categorical features of datasets into numerical features</b>; figuring out how best to handle categorical variables (like "Diet" and "Animal Class") is a big part of <b>data wrangling</b> for predictive modeling!
</div>

---

## 2. <u>Pandas Library</u><a id='pd'></a>

Now for the exciting part! Up to this point we have been fabricating data in the notebook to serve as our examples. With the introduction of the <b><a href='https://pandas.pydata.org/docs/user_guide/index.html'>Pandas Library</a></b>, we can **import** real data files into Jupyter Notebooks to explore. Let's do that now!

<br>

**Kaggle: our data source**  
We will use the <a href='https://www.kaggle.com/datasets/uciml/breast-cancer-wisconsin-data'>"Breast Cancer Wisconsin (Diagnostic) Dataset"</a> from **Kaggle**. Kaggle is a data science competition platform that hosts datathons, publish datasets, and support an online community of data scientists. Anyone is able to download the cleaned, published datasets to explore from the site and have access to an abundance of resources - from **data dictionaries** that detail data contents, to notebooks and code that other users of the data have posted. It's a great place to find interesting problems to explore and learn from others who have done/are doing the same.

<br>

**Pandas**  
Pandas is the standard tool for working with **dataframes**. A dataframe is a data structure that that represents data in a 2-dimensional table of rows and columns. We've seen a couple of examples of dataframes already, in the section on standard deviations, and just now in the matrix section. They are very useful for exploratory data analysis, data cleaning, and processing before turning them into matrices to be fed into machine learning models.

<div class="alert alert-info"><span style='color:#4169E1'>We've already imported the <code>pandas</code> library, but <b>let's do that again here:</b></span></div>

In [2]:
import pandas as pd

---

### 2.1. Reading in the Data<a id='import'></a>

Pandas allows us to easily "read in" data from a downloaded csv (comma separated values) file and save it as a variable in the Jupyter Notebook, with the `pd.read_csv()` function. It can take many different arguments depending on the desired specifications, but we can just accept the default for the optional parameters. The only required parameter is `filepath_or_buffer`, which asks for the **file path**, or location, of the data file on your computer so that it can find it and turn it into a Pandas dataframe. There are 2 ways to specify the file path:


<br><i class="fa fa-thumb-tack" style="font-size:16px;"></i>&nbsp; **Absolute File Path:** 

All of your files on the computer have a file path. If you go to the location of any file on your File Explorer, you can find its absolute file path by clicking the address bar at the top of the window. You'll see something like:

<code>C:/Users/username/folder/data_folder/filename.csv</code>

When you have all of the information needed to locate the file, all the way to the very first layer of folders, you have an <u>absolute</u> file path.


<br><i class="fa fa-thumb-tack" style="font-size:16px;"></i>&nbsp; **Relative File Path:**

Your Jupyter notebook (ipynb) file that you are working on has a path, too. When you navigate from one folder to the next on the File Explorer, you often start at a file location (let's call that location A), back out of that folder, enter into another folder, and access the file in this new location (location B). We can do something similar with file paths, by specifying the path of location B **relative to** the location of A.

Let's say this Jupyter notebook is found in location A, whose absolute path is:  
<code>C:/Users/username/folder/myNotebook.ipynb</code>

So, this is the **current directory**, or the location we are starting from:   
<code>C:/Users/username/folder/</code>

From here, we want to get to location B:  
<code>C:/Users/username/folder/data_folder/filename.csv</code>

<div class="alert alert-warning col-md-5 align=center"><b>To do this, we can specify the relative path:</b><br>  
<code>./data_folder/filename.csv</code></div> 

<br><br><br><br>

**Notation:**  
- The **`.`** in the relative path indicates we are **staying in the same, current directory**. Since the `data_folder` that contains the desired file is **inside** the current directory we started out in, we indicate that it is from here that we then move into another folder, or identify a file to point to.
<br>

- The **`..`** indicates we need to **back out of the current folder**. It's the equivalent of clicking the back button on File Explorer.  
>**Example**:  <br>
>We can back out of multiple folders - say there is another file in this location we want to get to:
>`C:/Users/another_user/theirFile.csv` 
>
>We can access this from location A with the relative path:  
>`../../another_user/theirFile.csv`


Once we have the path, all we need to do it put it in string form, an input it as an argument!

<div class="alert alert-info"><span style='color:#4169E1'><b>Run the code below</b> to read in our first dataset!</span></div>


In [3]:
# using relative path!
df = pd.read_csv("./data/data.csv")
df

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0      842302         M        17.99         10.38          122.80     1001.0   
1      842517         M        20.57         17.77          132.90     1326.0   
2    84300903         M        19.69         21.25          130.00     1203.0   
3    84348301         M        11.42         20.38           77.58      386.1   
4    84358402         M        20.29         14.34          135.10     1297.0   
..        ...       ...          ...           ...             ...        ...   
564    926424         M        21.56         22.39          142.00     1479.0   
565    926682         M        20.13         28.25          131.20     1261.0   
566    926954         M        16.60         28.08          108.30      858.1   
567    927241         M        20.60         29.33          140.10     1265.0   
568     92751         B         7.76         24.54           47.92      181.0   

     smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0            0.11840           0.27760         0.30010              0.14710   
1            0.08474           0.07864         0.08690              0.07017   
2            0.10960           0.15990         0.19740              0.12790   
3            0.14250           0.28390         0.24140              0.10520   
4            0.10030           0.13280         0.19800              0.10430   
..               ...               ...             ...                  ...   
564          0.11100           0.11590         0.24390              0.13890   
565          0.09780           0.10340         0.14400              0.09791   
566          0.08455           0.10230         0.09251              0.05302   
567          0.11780           0.27700         0.35140              0.15200   
568          0.05263           0.04362         0.00000              0.00000   

     ...  radius_worst  texture_worst  perimeter_worst  area_worst  \
0    ...        25.380          17.33           184.60      2019.0   
1    ...        24.990          23.41           158.80      1956.0   
2    ...        23.570          25.53           152.50      1709.0   
3    ...        14.910          26.50            98.87       567.7   
4    ...        22.540          16.67           152.20      1575.0   
..   ...           ...            ...              ...         ...   
564  ...        25.450          26.40           166.10      2027.0   
565  ...        23.690          38.25           155.00      1731.0   
566  ...        18.980          34.12           126.70      1124.0   
567  ...        25.740          39.42           184.60      1821.0   
568  ...         9.456          30.37            59.16       268.6   

     smoothness_worst  compactness_worst  concavity_worst  \
0             0.16220            0.66560           0.7119   
1             0.12380            0.18660           0.2416   
2             0.14440            0.42450           0.4504   
3             0.20980            0.86630           0.6869   
4             0.13740            0.20500           0.4000   
..                ...                ...              ...   
564           0.14100            0.21130           0.4107   
565           0.11660            0.19220           0.3215   
566           0.11390            0.30940           0.3403   
567           0.16500            0.86810           0.9387   
568           0.08996            0.06444           0.0000   

     concave points_worst  symmetry_worst  fractal_dimension_worst  
0                  0.2654          0.4601                  0.11890  
1                  0.1860          0.2750                  0.08902  
2                  0.2430          0.3613                  0.08758  
3                  0.2575          0.6638                  0.17300  
4                  0.1625          0.2364                  0.07678  
..                    ...             ...                      ...  
564                0.2216       

This dataset captures measurements and characteristics of breast mass (e.g. mass radius, smoothness, symmery) and the actual diagnosis of the mass. The challenge here would be to predict the diagnosis from the features of the mass. The purpose of this section is to introduce data manipulation using Pandas dataframes and series so we will not be tackling the challenge in this tutorial, but the notebooks uploaded on the <a href='https://www.kaggle.com/datasets/uciml/breast-cancer-wisconsin-data'>Kaggle page</a> would be a great place to see what other people have done with this dataset!

---

### 2.2. Exploring the Pandas DataFrame<a id='df'></a>

The Pandas DataFrame data structure allows us to easily access both the rows (records) **and** columns (features). It can be created in many different ways: from scratch, from a dictionary of values, from a matrix, from reading in a dataset, etc.


<br><i class="fa fa-thumb-tack" style="font-size:16px;"></i>&nbsp;**From scratch**:  

Below is an empty DataFrame object - it has no column or row yet. Run the code to see what it looks like:

In [82]:
df_fromScratch = pd.DataFrame()
df_fromScratch

Empty DataFrame
Columns: []
Index: []

We can add a column to the dataframe in the same way that we can add new key-value pairs to dictionaries:

In [83]:
df_fromScratch['first_column'] = np.arange(1, 10)
df_fromScratch['second_column']= np.arange(9, 0, -1)
df_fromScratch

first_column  second_column
0             1              9
1             2              8
2             3              7
3             4              6
4             5              5
5             6              4
6             7              3
7             8              2
8             9              1

However, once one column of a certain **length** is added to a dataframe, all other new columns must be of the same length:

In [84]:
# this will throw an error
df_fromScratch['short_column'] = np.array([7, 8, 9])

ValueError: Length of values does not match length of index

<br><i class="fa fa-thumb-tack" style="font-size:16px;"></i>&nbsp; **From a dictionary:**  

We can also convert a dictionary of data into a Pandas DataFrame, as long as the the number of elements captured in each dictionary value is the same:

In [85]:
# animal dictionary from earlier
pd.DataFrame(animal_dict)

Opposable Thumbs Class of Animal     Diet Tail Length  Number of Legs  \
0              True          Mammal  Bananas        long               2   
1             False       Anthropod  Insects        none               8   
2             False            Bird     Fish       short               2   

   Flies  
0  False  
1  False  
2  False

<br><i class="fa fa-thumb-tack" style="font-size:16px;"></i>&nbsp; **From a matrix:**  

..and same with matrices. Since a matrix does not have a name value like dictionaries do, we can include an argument to specify the column names:

In [86]:
columnNames = ['Opposable Thumbs', 'Class of Animal', 'Diet', 'Tail Length', 'Number of Legs', 'Flies']
pd.DataFrame(animal_matrix, columns = columnNames)

Opposable Thumbs Class of Animal     Diet Tail Length  Number of Legs  \
0              True          Mammal  Bananas        long               2   
1             False       Anthropod  Insects        none               9   
2             False            Bird     Fish       short               2   

   Flies  
0  False  
1  False  
2  False

<br><i class="fa fa-thumb-tack" style="font-size:16px;"></i>&nbsp; **Exploring data contents:**  

Let's explore the Pandas capabilities using breast cancer data we read in earlier. The first step we'd want to take when exploring a dataset is to undertand what the dataset contains. The DataFrame object has many attributes to help us with this task:

1. Identify the number of rows (records) and columns (features) in the data
2. Get info on the column names, their position on the dataframe, how many non-**null\*** values there are in each feature, and what the data type of each feature is
3. Get a list of the columns in the dataset
4. Create a table of summary statistics on all numeric features



In [87]:
# 1. find the number of rows and columns (row, col) in the dataset
df.shape

(569, 32)

In [4]:
# 2. summary of features
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 32 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   id                       569 non-null    int64  
 1   diagnosis                569 non-null    object 
 2   radius_mean              569 non-null    float64
 3   texture_mean             569 non-null    float64
 4   perimeter_mean           569 non-null    float64
 5   area_mean                569 non-null    float64
 6   smoothness_mean          569 non-null    float64
 7   compactness_mean         569 non-null    float64
 8   concavity_mean           569 non-null    float64
 9   concave points_mean      569 non-null    float64
 10  symmetry_mean            569 non-null    float64
 11  fractal_dimension_mean   569 non-null    float64
 12  radius_se                569 non-null    float64
 13  texture_se               569 non-null    float64
 14  perimeter_se             5

In [89]:
# 3. list of column names
df.columns

Index(['id', 'diagnosis', 'radius_mean', 'texture_mean', 'perimeter_mean',
       'area_mean', 'smoothness_mean', 'compactness_mean', 'concavity_mean',
       'concave points_mean', 'symmetry_mean', 'fractal_dimension_mean',
       'radius_se', 'texture_se', 'perimeter_se', 'area_se', 'smoothness_se',
       'compactness_se', 'concavity_se', 'concave points_se', 'symmetry_se',
       'fractal_dimension_se', 'radius_worst', 'texture_worst',
       'perimeter_worst', 'area_worst', 'smoothness_worst',
       'compactness_worst', 'concavity_worst', 'concave points_worst',
       'symmetry_worst', 'fractal_dimension_worst'],
      dtype='object')

In [90]:
# 4. summary statistics of numeric features
df.describe()

id  radius_mean  texture_mean  perimeter_mean    area_mean  \
count  5.690000e+02   569.000000    569.000000      569.000000   569.000000   
mean   3.037183e+07    14.127292     19.289649       91.969033   654.889104   
std    1.250206e+08     3.524049      4.301036       24.298981   351.914129   
min    8.670000e+03     6.981000      9.710000       43.790000   143.500000   
25%    8.692180e+05    11.700000     16.170000       75.170000   420.300000   
50%    9.060240e+05    13.370000     18.840000       86.240000   551.100000   
75%    8.813129e+06    15.780000     21.800000      104.100000   782.700000   
max    9.113205e+08    28.110000     39.280000      188.500000  2501.000000   

       smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
count       569.000000        569.000000      569.000000           569.000000   
mean          0.096360          0.104341        0.088799             0.048919   
std           0.014064          0.052813        0.079720             0.038803   
min           0.052630          0.019380        0.000000             0.000000   
25%           0.086370          0.064920        0.029560             0.020310   
50%           0.095870          0.092630        0.061540             0.033500   
75%           0.105300          0.130400        0.130700             0.074000   
max           0.163400          0.345400        0.426800             0.201200   

       symmetry_mean  ...  radius_worst  texture_worst  perimeter_worst  \
count     569.000000  ...    569.000000     569.000000       569.000000   
mean        0.181162  ...     16.269190      25.677223       107.261213   
std         0.027414  ...      4.833242       6.146258        33.602542   
min         0.106000  ...      7.930000      12.020000        50.410000   
25%         0.161900  ...     13.010000      21.080000        84.110000   
50%         0.179200  ...     14.970000      25.410000        97.660000   
75%         0.195700  ...     18.790000      29.720000       125.400000   
max         0.304000  ...     36.040000      49.540000       251.200000   

        area_worst  smoothness_worst  compactness_worst  concavity_worst  \
count   569.000000        569.000000         569.000000       569.000000   
mean    880.583128          0.132369           0.254265         0.272188   
std     569.356993          0.022832           0.157336         0.208624   
min     185.200000          0.071170           0.027290         0.000000   
25%     515.300000          0.116600           0.147200         0.114500   
50%     686.500000          0.131300           0.211900         0.226700   
75%    1084.000000          0.146000           0.339100         0.382900   
max    4254.000000          0.222600           1.058000         1.252000   

       concave points_worst  symmetry_worst  fractal_dimension_worst  
count            569.000000      569.000000               569.000000  
mean               0.114606        0.290076                 0.083946  
std                0.065732        0.061867                 0.018061  
min                0.000000        0.156500                 0.055040  
25%                0.064930        0.250400                 0.071460  
50%                0.099930        0.282200                 0.080040  
75%                0.161400        0.317900                 0.092080  
max                0.291000        0.663800                 0.207500  

[8 rows x 31 columns]

<div class="alert alert-warning">
<i class="fa fa-info-circle" style="font-size:22px;color:orange"></i> &nbsp;A <b>null</b> or <b>nan</b> value represents an unknown or missing value in the data - it is an empty entry. If a feature is riddled with missing values, we may need to drop the feature from the investigation since it may not capture enough valid data, or the valid data it does capture may be biased. If a feature has some missing values but still captures valuable information, we will want to clean the data so to replace these values with something more interpretable. We won't touch on this here, but you can find a <a href='https://pandas.pydata.org/docs/user_guide/missing_data.html'>guide on how to work with missing data using Pandas</a> in their documentation.
</div>



<br><i class="fa fa-thumb-tack" style="font-size:16px;"></i>&nbsp;**Selecting a DataFrame feature - Series**:  <br>

We know from the `.info` output above that there is only one non-numeric field in the dataset, and that is the target variable - the diagnosis. Let's understand this target variable better.

We can select a feature from the DataFrame in a similar way to how we would get the value of a dictionary - by indexing the dataframe by the column name:

In [91]:
# get the data values of `diagnosis` column
df['diagnosis']

0      M
1      M
2      M
3      M
4      M
      ..
564    M
565    M
566    M
567    M
568    B
Name: diagnosis, Length: 569, dtype: object

The extracted column is stored in a Pandas data structure called a **Pandas Series**. 

<i class="fa fa-book" style="font-size:20px;"></i> &nbsp;**Definition:**
<div class="alert alert-success">
    A <b>Series</b> is a Pandas data structure that behaves very similarly to NumPy arrays and will be a valid argument to most NumPy functions. Series are also similar to dictionaries, in that its values can have index labels and be indexed by these labels.
</div>

For instance:

In [92]:
# this is an array
array_a = np.arange(1, 6)

# this is a Series, with non-numeric index labels, and a name
series_a = pd.Series(array_a, index=['a', 'b', 'c', 'd', 'e'], name="Series_A")

print("array: ", array_a)
print("\nSeries: ")
print(series_a)

array:  [1 2 3 4 5]

Series: 
a    1
b    2
c    3
d    4
e    5
Name: Series_A, dtype: int32


`series_a` has non-numeric indices. If I want to extract a value from the structure, I can index using its positional index (like an array), or using its label index (like a dictionary):

In [93]:
# extracting value like an array
series_a[1]

2

In [94]:
# extracting value like a dictionary 
series_a['b']

2

A Series can also have a `name` attribute, which is how Pandas knows to name the dataframe when a Series object is turned into a dataframe:

In [95]:
series_a.to_frame()

Series_A
a         1
b         2
c         3
d         4
e         5

**Now back to our data.** If we were to predict the diagnosis based on the cancer mass attributes, it would be good to know how many categories of diagnoses there may be. We want to find the unique values of the variable.

Like the DataFrame object, the Pandas Series object also has many useful attributes. Let's use a couple of them here to better understand the field:


In [96]:
# Find all unique values of the field
df['diagnosis'].unique()

array(['M', 'B'], dtype=object)

There are only 2 possible values for the `diagnosis` variable - malignant (`M`) and benign (`B`). Use the `.value_counts()` method to count how many of each are in the dataset:

In [5]:
df['diagnosis'].value_counts()

B    357
M    212
Name: diagnosis, dtype: int64

### 2.3 Selecting Rows & Columns<a id="loc"></a>

What if we wanted to create subsets of our data, without pulling them out one by one into Pandas Series objects? We will often want to select a set of features (columns) to keep, or filter for data records (rows) that meet specific criteria. There are a number of ways to accomplish this:

<br><i class="fa fa-thumb-tack" style="font-size:16px;"></i>&nbsp;**Creating a dataset with fewer selected features:**  <br>

Often times we want to investigate just a couple of fields from the data. In these cases, we may want to create a smaller dataset for greater efficiency and run times. We can select fields to keep in a few ways:

**1. Double square brackets `[]`**  
We can index the dataframe with a list of column names to create a dataset with just those columns (but with all the rows).

In [98]:
df[['diagnosis', 'area_mean']]

diagnosis  area_mean
0           M     1001.0
1           M     1326.0
2           M     1203.0
3           M      386.1
4           M     1297.0
..        ...        ...
564         M     1479.0
565         M     1261.0
566         M      858.1
567         M     1265.0
568         B      181.0

[569 rows x 2 columns]

**2. `.loc[]` attribute**   
We can do the same using the `.loc` attribute. This method allows us to specify the column names to keep **and** filter the rows at the same time.

Just as we could slice (extract specific ranges of) sequences based their positional indices, we can slice the data rows and data columns by their index labels. 

The `.loc[]` method takes two ranges. The range for rows is specified first, and the range for columns second: 

df.loc\[ <span style='color:green'>startRowLabel<b> : </b>endRowLabel</span>, <span style='color : navy'>startColName<b> : </b>endColName</span> \]

In [99]:
# grabs all records, and all columns positioned between and including `diagnosis` and `area_mean`
df.loc[:, "diagnosis":"area_mean"]

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean
0           M        17.99         10.38          122.80     1001.0
1           M        20.57         17.77          132.90     1326.0
2           M        19.69         21.25          130.00     1203.0
3           M        11.42         20.38           77.58      386.1
4           M        20.29         14.34          135.10     1297.0
..        ...          ...           ...             ...        ...
564         M        21.56         22.39          142.00     1479.0
565         M        20.13         28.25          131.20     1261.0
566         M        16.60         28.08          108.30      858.1
567         M        20.60         29.33          140.10     1265.0
568         B         7.76         24.54           47.92      181.0

[569 rows x 5 columns]

In [100]:
# if we just want the 2 columns and not the columns in between, we leverage the double-bracket
df.loc[:, ["diagnosis","area_mean"]]

diagnosis  area_mean
0           M     1001.0
1           M     1326.0
2           M     1203.0
3           M      386.1
4           M     1297.0
..        ...        ...
564         M     1479.0
565         M     1261.0
566         M      858.1
567         M     1265.0
568         B      181.0

[569 rows x 2 columns]

**3. `.iloc[]` - attribute**  
This is very similar to `.loc[]`, but instead of using row and column labels, we specify index positions instead. We can see below that `diagnosis` is found at index `1`, and `area_mean` at index `5`. So, if we specify the range `1:6`, we should get the same table as before.

*Remember that, when slicing with indices, the `stop` value in the `start`:`stop` range is excluded from the selection.*

In [101]:
# run to see column positions
df.columns

Index(['id', 'diagnosis', 'radius_mean', 'texture_mean', 'perimeter_mean',
       'area_mean', 'smoothness_mean', 'compactness_mean', 'concavity_mean',
       'concave points_mean', 'symmetry_mean', 'fractal_dimension_mean',
       'radius_se', 'texture_se', 'perimeter_se', 'area_se', 'smoothness_se',
       'compactness_se', 'concavity_se', 'concave points_se', 'symmetry_se',
       'fractal_dimension_se', 'radius_worst', 'texture_worst',
       'perimeter_worst', 'area_worst', 'smoothness_worst',
       'compactness_worst', 'concavity_worst', 'concave points_worst',
       'symmetry_worst', 'fractal_dimension_worst'],
      dtype='object')

In [102]:
# slicing using index positions
df.iloc[:, 1:6]

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean
0           M        17.99         10.38          122.80     1001.0
1           M        20.57         17.77          132.90     1326.0
2           M        19.69         21.25          130.00     1203.0
3           M        11.42         20.38           77.58      386.1
4           M        20.29         14.34          135.10     1297.0
..        ...          ...           ...             ...        ...
564         M        21.56         22.39          142.00     1479.0
565         M        20.13         28.25          131.20     1261.0
566         M        16.60         28.08          108.30      858.1
567         M        20.60         29.33          140.10     1265.0
568         B         7.76         24.54           47.92      181.0

[569 rows x 5 columns]

In [103]:
df.iloc[:, [1, 5]]

diagnosis  area_mean
0           M     1001.0
1           M     1326.0
2           M     1203.0
3           M      386.1
4           M     1297.0
..        ...        ...
564         M     1479.0
565         M     1261.0
566         M      858.1
567         M     1265.0
568         B      181.0

[569 rows x 2 columns]

<br><i class="fa fa-thumb-tack" style="font-size:16px;"></i>&nbsp;**Slicing and Filtering Dataset Records:**  <br>

Just as we can create data with subsets of columns, we can create data with subsets of rows.

**1. Regular indexing**  
When we specify a range of integer values, DataFrames know to slice the rows:

In [104]:
# keep first 100 records
df[:100]

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0     842302         M       17.990         10.38          122.80     1001.0   
1     842517         M       20.570         17.77          132.90     1326.0   
2   84300903         M       19.690         21.25          130.00     1203.0   
3   84348301         M       11.420         20.38           77.58      386.1   
4   84358402         M       20.290         14.34          135.10     1297.0   
..       ...       ...          ...           ...             ...        ...   
95     86208         M       20.260         23.03          132.40     1264.0   
96     86211         B       12.180         17.84           77.79      451.1   
97    862261         B        9.787         19.94           62.11      294.5   
98    862485         B       11.600         12.84           74.34      412.6   
99    862548         M       14.420         19.77           94.48      642.5   

    smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0           0.11840           0.27760        0.300100             0.147100   
1           0.08474           0.07864        0.086900             0.070170   
2           0.10960           0.15990        0.197400             0.127900   
3           0.14250           0.28390        0.241400             0.105200   
4           0.10030           0.13280        0.198000             0.104300   
..              ...               ...             ...                  ...   
95          0.09078           0.13130        0.146500             0.086830   
96          0.10450           0.07057        0.024900             0.029410   
97          0.10240           0.05301        0.006829             0.007937   
98          0.08983           0.07525        0.041960             0.033500   
99          0.09752           0.11410        0.093880             0.058390   

    ...  radius_worst  texture_worst  perimeter_worst  area_worst  \
0   ...         25.38          17.33           184.60      2019.0   
1   ...         24.99          23.41           158.80      1956.0   
2   ...         23.57          25.53           152.50      1709.0   
3   ...         14.91          26.50            98.87       567.7   
4   ...         22.54          16.67           152.20      1575.0   
..  ...           ...            ...              ...         ...   
95  ...         24.22          31.59           156.10      1750.0   
96  ...         12.83          20.92            82.14       495.2   
97  ...         10.92          26.29            68.81       366.1   
98  ...         13.06          17.16            82.96       512.5   
99  ...         16.33          30.86           109.50       826.4   

    smoothness_worst  compactness_worst  concavity_worst  \
0             0.1622            0.66560          0.71190   
1             0.1238            0.18660          0.24160   
2             0.1444            0.42450          0.45040   
3             0.2098            0.86630          0.68690   
4             0.1374            0.20500          0.40000   
..               ...                ...              ...   
95            0.1190            0.35390          0.40980   
96            0.1140            0.09358          0.04980   
97            0.1316            0.09473          0.02049   
98            0.1431            0.18510          0.19220   
99            0.1431            0.30260          0.31940   

    concave points_worst  symmetry_worst  fractal_dimension_worst  
0                0.26540          0.4601                  0.11890  
1                0.18600          0.2750                  0.08902  
2                0.24300          0.3613                  0.08758  
3                0.25750          0.6638                  0.17300  
4                0.16250          0.2364                  0.07678  
..                   ...             ...                      ...  
95               0.15730          0.3689                  0.08368  
96               0

**2. Filtering by criteria**  
We can also filter by a criteria in the data. For instance, what if we only wanted to check out the distributions of features for masses that are known to be "benign"? We would create what we call a **mask**, and apply it to the dataset, like this:


In [105]:
# applying a mask to the dataset
# to only keep records that are benign
df[df['diagnosis']=='B']

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
19    8510426         B       13.540         14.36           87.46      566.3   
20    8510653         B       13.080         15.71           85.63      520.0   
21    8510824         B        9.504         12.44           60.34      273.9   
37     854941         B       13.030         18.42           82.61      523.8   
46   85713702         B        8.196         16.84           51.71      201.9   
..        ...       ...          ...           ...             ...        ...   
558    925277         B       14.590         22.68           96.39      657.1   
559    925291         B       11.510         23.93           74.52      403.5   
560    925292         B       14.050         27.15           91.38      600.4   
561    925311         B       11.200         29.37           70.67      386.0   
568     92751         B        7.760         24.54           47.92      181.0   

     smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
19           0.09779           0.08129         0.06664             0.047810   
20           0.10750           0.12700         0.04568             0.031100   
21           0.10240           0.06492         0.02956             0.020760   
37           0.08983           0.03766         0.02562             0.029230   
46           0.08600           0.05943         0.01588             0.005917   
..               ...               ...             ...                  ...   
558          0.08473           0.13300         0.10290             0.037360   
559          0.09261           0.10210         0.11120             0.041050   
560          0.09929           0.11260         0.04462             0.043040   
561          0.07449           0.03558         0.00000             0.000000   
568          0.05263           0.04362         0.00000             0.000000   

     ...  radius_worst  texture_worst  perimeter_worst  area_worst  \
19   ...        15.110          19.26            99.70       711.2   
20   ...        14.500          20.49            96.09       630.5   
21   ...        10.230          15.66            65.13       314.9   
37   ...        13.300          22.81            84.46       545.9   
46   ...         8.964          21.96            57.26       242.2   
..   ...           ...            ...              ...         ...   
558  ...        15.480          27.27           105.90       733.5   
559  ...        12.480          37.16            82.28       474.2   
560  ...        15.300          33.17           100.20       706.7   
561  ...        11.920          38.30            75.19       439.6   
568  ...         9.456          30.37            59.16       268.6   

     smoothness_worst  compactness_worst  concavity_worst  \
19            0.14400            0.17730          0.23900   
20            0.13120            0.27760          0.18900   
21            0.13240            0.11480          0.08867   
37            0.09701            0.04619          0.04833   
46            0.12970            0.13570          0.06880   
..                ...                ...              ...   
558           0.10260            0.31710          0.36620   
559           0.12980            0.25170          0.36300   
560           0.12410            0.22640          0.13260   
561           0.09267            0.05494          0.00000   
568           0.08996            0.06444          0.00000   

     concave points_worst  symmetry_worst  fractal_dimension_worst  
19                0.12880          0.2977                  0.07259  
20                0.07283          0.3184                  0.08183  
21                0.06227          0.2450                  0.07773  
37                0.05013          0.1987                  0.06169  
46                0.02564          0.3105                  0.07409  
..                    ...             ...                      ...  
558               0.11050       

Recall that a Series acts very much like a NumPy array. This mean that the expression `df['diagnosis']=='B'` would create a long array of `True` and `False`, depending on whether the element in the `diagnosis` field is `=='B'` or not. This sequence of boolean values acts as a mask on the dataset - the DataFrame knows only to keep records that contains a `True` value from the mask.

In [106]:
# mask
df['diagnosis']=='B'

0      False
1      False
2      False
3      False
4      False
       ...  
564    False
565    False
566    False
567    False
568     True
Name: diagnosis, Length: 569, dtype: bool

**2. `.loc[]` and `.iloc[]` attributes** 

The `.loc[]` attribute supports slicing and filtering rows, as well:

In [107]:
# slicing rows using index values (which in this case is same as index positions)
df.loc[:100]

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0      842302         M       17.990         10.38          122.80     1001.0   
1      842517         M       20.570         17.77          132.90     1326.0   
2    84300903         M       19.690         21.25          130.00     1203.0   
3    84348301         M       11.420         20.38           77.58      386.1   
4    84358402         M       20.290         14.34          135.10     1297.0   
..        ...       ...          ...           ...             ...        ...   
96      86211         B       12.180         17.84           77.79      451.1   
97     862261         B        9.787         19.94           62.11      294.5   
98     862485         B       11.600         12.84           74.34      412.6   
99     862548         M       14.420         19.77           94.48      642.5   
100    862717         M       13.610         24.98           88.05      582.7   

     smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0            0.11840           0.27760        0.300100             0.147100   
1            0.08474           0.07864        0.086900             0.070170   
2            0.10960           0.15990        0.197400             0.127900   
3            0.14250           0.28390        0.241400             0.105200   
4            0.10030           0.13280        0.198000             0.104300   
..               ...               ...             ...                  ...   
96           0.10450           0.07057        0.024900             0.029410   
97           0.10240           0.05301        0.006829             0.007937   
98           0.08983           0.07525        0.041960             0.033500   
99           0.09752           0.11410        0.093880             0.058390   
100          0.09488           0.08511        0.086250             0.044890   

     ...  radius_worst  texture_worst  perimeter_worst  area_worst  \
0    ...         25.38          17.33           184.60      2019.0   
1    ...         24.99          23.41           158.80      1956.0   
2    ...         23.57          25.53           152.50      1709.0   
3    ...         14.91          26.50            98.87       567.7   
4    ...         22.54          16.67           152.20      1575.0   
..   ...           ...            ...              ...         ...   
96   ...         12.83          20.92            82.14       495.2   
97   ...         10.92          26.29            68.81       366.1   
98   ...         13.06          17.16            82.96       512.5   
99   ...         16.33          30.86           109.50       826.4   
100  ...         16.99          35.27           108.60       906.5   

     smoothness_worst  compactness_worst  concavity_worst  \
0              0.1622            0.66560          0.71190   
1              0.1238            0.18660          0.24160   
2              0.1444            0.42450          0.45040   
3              0.2098            0.86630          0.68690   
4              0.1374            0.20500          0.40000   
..                ...                ...              ...   
96             0.1140            0.09358          0.04980   
97             0.1316            0.09473          0.02049   
98             0.1431            0.18510          0.19220   
99             0.1431            0.30260          0.31940   
100            0.1265            0.19430          0.31690   

     concave points_worst  symmetry_worst  fractal_dimension_worst  
0                 0.26540          0.4601                  0.11890  
1                 0.18600          0.2750                  0.08902  
2                 0.24300          0.3613                  0.08758  
3                 0.25750          0.6638                  0.17300  
4                 0.16250          0.2364                  0.07678  
..                    ...             ...                      ...  
96                0.05882       

The nice thing about `.loc[]` is that it allows you to filter or slice for rows and columns at the same time:

In [108]:
# filtering for benign records
df.loc[df['diagnosis']=='B', 'diagnosis':'area_mean']

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean
19          B       13.540         14.36           87.46      566.3
20          B       13.080         15.71           85.63      520.0
21          B        9.504         12.44           60.34      273.9
37          B       13.030         18.42           82.61      523.8
46          B        8.196         16.84           51.71      201.9
..        ...          ...           ...             ...        ...
558         B       14.590         22.68           96.39      657.1
559         B       11.510         23.93           74.52      403.5
560         B       14.050         27.15           91.38      600.4
561         B       11.200         29.37           70.67      386.0
568         B        7.760         24.54           47.92      181.0

[357 rows x 5 columns]

We can also slice rows and columns simultaneously with the `.iloc[]` attribute:

In [109]:
df.iloc[:100, [1, 5]]

diagnosis  area_mean
0          M     1001.0
1          M     1326.0
2          M     1203.0
3          M      386.1
4          M     1297.0
..       ...        ...
95         M     1264.0
96         B      451.1
97         B      294.5
98         B      412.6
99         M      642.5

[100 rows x 2 columns]

### 2.4 Feature Engineering & Applying Functions<a id="apply"></a>

Often, we will want to engineer new features from existing ones or transform features in the dataset. We'll approach this in a couple of ways:

1. Computing a new sequence of data with existing ones, and assigning it as a new column
2. Using the `.apply()` DataFrame method

<br><i class="fa fa-thumb-tack" style="font-size:16px;"></i>&nbsp; **1. Creating a new potential feature**</br>

Looking at the available fields, it looks like there might be an opportunity to approximate how *irregularly shaped* a mass may have been. In particular, we are interested in these data fields:

In [6]:
df_new = df.loc[:, ['area_worst', 'radius_worst', 'perimeter_worst', 'symmetry_worst', 'diagnosis']]
df_new

area_worst  radius_worst  perimeter_worst  symmetry_worst diagnosis
0        2019.0        25.380           184.60          0.4601         M
1        1956.0        24.990           158.80          0.2750         M
2        1709.0        23.570           152.50          0.3613         M
3         567.7        14.910            98.87          0.6638         M
4        1575.0        22.540           152.20          0.2364         M
..          ...           ...              ...             ...       ...
564      2027.0        25.450           166.10          0.2060         M
565      1731.0        23.690           155.00          0.2572         M
566      1124.0        18.980           126.70          0.2218         M
567      1821.0        25.740           184.60          0.4087         M
568       268.6         9.456            59.16          0.2871         B

[569 rows x 5 columns]

We know that the area of a circle is found by the equation $A = \pi r^2$, and that the circumference of a circle is given by $C = 2 \pi r$. If we make the assumption that the mass is **not** irrecgularly shaped, i.e. the mass has a circular shape, then the measured perimeter of the mass and the calculated circumference should in theory be pretty similar. If the perimeter is larger than the circumference by a lot, that may be a good indicator that there is irregulary in the shape, which may be a good predictor of a malignant mass.

<div class="alert alert-info"><span style='color:#4169E1'>Let's <b>calculate the circumference of the mass</b> given its measured area and radius, and create a new field that captures the <b>ratio of the calculated circumference to the measured perimeter:</b></span></div>



In [7]:
# Series behave like NumPy arrays - the same rules of arithmetic operations apply here

# C = 2*A/r : circumference = 2 x area / radius
circumference = 2*df_new['area_worst']/df_new['radius_worst']

# creating the new ratio field
df_new['ratio_CtoP'] = circumference / df_new['perimeter_worst']

df_new

area_worst  radius_worst  perimeter_worst  symmetry_worst diagnosis  \
0        2019.0        25.380           184.60          0.4601         M   
1        1956.0        24.990           158.80          0.2750         M   
2        1709.0        23.570           152.50          0.3613         M   
3         567.7        14.910            98.87          0.6638         M   
4        1575.0        22.540           152.20          0.2364         M   
..          ...           ...              ...             ...       ...   
564      2027.0        25.450           166.10          0.2060         M   
565      1731.0        23.690           155.00          0.2572         M   
566      1124.0        18.980           126.70          0.2218         M   
567      1821.0        25.740           184.60          0.4087         M   
568       268.6         9.456            59.16          0.2871         B   

     ratio_CtoP  
0      0.861872  
1      0.985785  
2      0.950917  
3      0.770206  
4      0.918210  
..          ...  
564    0.959017  
565    0.942823  
566    0.934810  
567    0.766478  
568    0.960286  

[569 rows x 6 columns]

Nice! We have engineered our first feature.

<br>

<br><i class="fa fa-thumb-tack" style="font-size:16px;"></i>&nbsp; **2. `apply()`**</br>  

`.apply` allows us to take a function and apply it to the Pandas series or dataframe. 

<div class="alert alert-info"><span style='color:#4169E1'>Let's <b>standardize the columns <code>area_worst</code>, <code>radius_worst</code>, and <code>perimeter_worst</code></b> by applying the function we had defined earlier:</span></div>


In [112]:
# check the docs for more details!
df.apply?

In [18]:
# function to standardize data
def standard_units(numbers_array):
    "Convert an array of numbers to standard units"
    return (numbers_array - np.mean(numbers_array))/np.std(numbers_array)

In [114]:
# applying the function to the 3 fields
df.loc[:, ['area_worst', 'radius_worst', 'perimeter_worst']].apply(standard_units)

area_worst  radius_worst  perimeter_worst
0      2.001237      1.886690         2.303601
1      1.890489      1.805927         1.535126
2      1.456285      1.511870         1.347475
3     -0.550021     -0.281464        -0.249939
4      1.220724      1.298575         1.338539
..          ...           ...              ...
564    2.015301      1.901185         1.752563
565    1.494959      1.536720         1.421940
566    0.427906      0.561361         0.579001
567    1.653171      1.961239         2.303601
568   -1.075813     -1.410893        -1.432735

[569 rows x 3 columns]

..so much more elegant than extracting each individual field as a Series, plugging them into the function, and setting each new output a as a new column in the dataset!

### 2.5 Data Aggregation<a id="group"></a>
The final concept we will cover is the concept of **grouped operations**. Grouping datasets allow us to efficiently compute and compare aggregations of data values conducted separately for each group of fields.

In the section above, we have just explored shape irregularity as a possible predictor of malignant vs. benign masses. One way to analyze whether we may be onto something is to compute the feature's summary statistic separately for the two groups, and see if we observe a notable difference. We can do this with the `.groupby()` method for Pandas DataFrames, which organizes the data into groups based on the values of the group-by variable, and computes an aggregation on the members of each group such that we are left with an aggregate value for each group. It takes the form:

`df.groupby(group_variable).aggregation()`


<div class="alert alert-info"><span style='color:#4169E1'>Let's <b>find the averages by diagnosis</b> of the new and existing features in the <code>df_new</code> data:</span></div>


In [115]:
df_new.groupby('diagnosis').mean()

area_worst  radius_worst  perimeter_worst  symmetry_worst  \
diagnosis                                                               
B           558.899440     13.379801        87.005938        0.270246   
M          1422.286321     21.134811       141.370330        0.323468   

           ratio_CtoP  
diagnosis              
B            0.940368  
M            0.916292

Recall that there are only two possible diagnoses: **B**enign, or **M**alignant. We had seen earlier from looking at its `.value_counts()` that there are 357 benign records and 212 malignant records. The `groupby` operation above is calculating the averages for each of the 5 features in the `df_new` dataset, for both the group of 357 benign records and, separately, 212 malignant records. If we were to filter the dataset for benign records, isolate the "area_worst" field, and calculate its mean, we would arrive at the value in the upper left-most cell:

In [17]:
# verify that this matches the mean of "area_worst" for the benign diagnostic group:
print("Benign cases:", df_new[df_new['diagnosis']=="B"]['area_worst'].mean())

Benign cases: 558.8994397759104


Aggregation allows us to quickly consolidate data by a specific category(ies). We can apply built-in aggregators (like `.mean()`) or user-defined aggregating functions, like below:

In [19]:
# mean of standard units should be very close to 0 
def meanOfStandardUnits(numbers_array):
    return np.mean(standard_units(numbers_array))

In [23]:
df_new.groupby('diagnosis').aggregate(meanOfStandardUnits)

area_worst  radius_worst  perimeter_worst  symmetry_worst  \
diagnosis                                                                
B         -2.711805e-16 -4.830248e-15    -3.185749e-15   -4.832736e-16   
M         -1.838152e-16  2.308426e-15     8.117197e-16    2.534660e-16   

             ratio_CtoP  
diagnosis                
B         -6.974813e-15  
M         -3.703536e-15


**Congratulations!** You've made it through the datathon tutorial notebooks. While there is a lot more to learn beyond what's covered in this tutorial when it comes to the art and science of working with data, you have begun to build a solid foundation from which you can dive into the world of data science. As long as you remain curious and leverage the many resources available to you (documentation sites, Kaggle community, Stack Overflow, WiDS workshops, etc.), you are bound to rapidly develop your data science repertoire. Good luck, and have fun!

---

#### Content adapted from:  
- Jupyter Notebook modules from the [UC Berkeley Data Science Modules Program](https://ds-modules.github.io/DS-Modules/) licensed under [CC BY-NC 4.0](https://creativecommons.org/licenses/by-nc/4.0/)
    - [Data 8X Public Materials for 2022](https://github.com/ds-modules/materials-x22/) by Sean Morris
- [Composing Programs](https://www.composingprograms.com/) by John DeNero based on the textbook [Structure and Interpretation of Computer Programs](https://mitpress.mit.edu/9780262510875/structure-and-interpretation-of-computer-programs/) by Harold Abelson and Gerald Jay Sussman, licensed under [CC BY-SA 3.0](https://creativecommons.org/licenses/by-sa/3.0/)  
